src/coins.cpp

```diff
@@ -1102,3 +1171,136 @@ CCoinsModifier::~CCoinsModifier()
+    if (!initialized) {
+        parentLatestSubtree = parentView->GetLatestSubtree(type);
+        initialized = true;
+    }
+
+    if (newSubtrees.size() > 0) {
+        // The latest subtree is in our cache.
+
+        libzcash::SubtreeIndex index;
+        if (parentLatestSubtree.has_value()) {
+            // The best subtree index is newSubtrees.size() larger than
+            // our parent view's subtree index.
+            index = parentLatestSubtree.value().index + newSubtrees.size();
+        } else {
+            // The parent view has no subtrees
+            index = newSubtrees.size() - 1;
+        }
+
+        auto lastSubtree = newSubtrees.back();
+        return libzcash::LatestSubtree(index, lastSubtree.root, lastSubtree.nHeight);
+    } else {
+        return parentLatestSubtree;
+    }
+}
+
+std::optional<libzcash::SubtreeData> SubtreeCache::GetSubtreeData(CCoinsView *parentView, libzcash::SubtreeIndex index) {
+    auto latestSubtree = GetLatestSubtree(parentView);
+
+    if (!latestSubtree.has_value() || latestSubtree.value().index < index) {
+        // This subtree isn't complete in our local view
+        return std::nullopt;
+    }
+
+    if (parentLatestSubtree.has_value() && (index <= parentLatestSubtree.value().index)) {
+        // This subtree is complete, but it's not in the local cache
+        return parentView->GetSubtreeData(type, index);
+    } else {
+        if (parentLatestSubtree.has_value()) {
+            // Get the index into our local `newSubtrees` that the subtree should
+            // be located.
+            auto localIndex = index - (parentLatestSubtree.value().index + 1);
+            assert(newSubtrees.size() > localIndex);
+            return newSubtrees[localIndex];
+        } else {
+            // The index we've been given is the index into our local `newSubtrees`
+            // since the parent view has no subtrees.
+            assert(newSubtrees.size() > index);
+            return newSubtrees[index];
+        }
+    }
+}
+
+void SubtreeCache::PushSubtree(CCoinsView *parentView, libzcash::SubtreeData subtree) {
+    GetLatestSubtree(parentView);
+
+    newSubtrees.push_back(subtree);
+}
+
+void SubtreeCache::PopSubtree(CCoinsView *parentView) {
+    GetLatestSubtree(parentView);
+
+    if (newSubtrees.empty()) {
+        // Try to pop from the parent view
+        if (parentLatestSubtree.has_value()) {
+            libzcash::SubtreeIndex parentIndex = parentLatestSubtree.value().index;
+
+            if (parentIndex == 0) {
+                // This pops the only subtree left in the parent view.
+                parentLatestSubtree = std::nullopt;
+            } else {
+                parentIndex -= 1;
+                auto newParent = parentView->GetSubtreeData(type, parentIndex);
+                if (!newParent.has_value()) {
+                    throw std::runtime_error("cache inconsistency; parent view does not have subtree");
+                }
```

Comment:
```suggestion
                assert(newParent.has_value());
```

Reply:
It's hard to thread the needle over whether some of these should be assertions and some should throw exceptions. Technically, the user of `CCoinsViewCache` can change the `base` stored within the cache arbitrarily, which they shouldn't do in most contexts unless they know what they're doing. Theoretically they could do this and break this function, which would make it less useful as an assertion and more useful as an exception. Whereas other places where I've added assertions _should_ be only tripped up if the code within the `SubtreeCache`/`CCoinsViewCache` is actually wrong somehow.